docs/rfcs/0011-customizable-store-dispatcher.md

```diff
@@ -0,0 +1,71 @@
+---
+layout: page
+title: Allow customisation of the Store dispatcher
+permalink: /rfc/0011-customizable-store-dispatcher
+---
+
+* Start date: 2024-01-29
+* RFC PR: [#5353](https://github.com/mozilla-mobile/firefox-android/pull/5353)
+
+## Summary
+
+In most applications of the `Store`, it is preferable to have reducers perform work on the main thread. This helps to simplify the levels of async tasks that need to be synchronised with other async tasks that are already on the main thread.
+
+## Motivation
+
+Android embedders use the main thread for UI, user-facing, or touch handling work. For example, notifying UI components when IO from storage layers have completed, an engine's tasks that can happen on a separate thread, or global-level state updates for different components to observe.
+
+When components dispatch actions to the store, these actions are performed on an independant single thread dispatcher in the `Store` to avoid overloading the main thread with heavy work that might be performed during the `reduce` or in a `Middleware`. In practice, these actions have been short and fast so they do not cause overhead. Work done in a `Middleware` that can be slow, like I/O,  is put onto separate Dispatchers. The performance optimization to switch to a `Store` thread, requires that components which are always run on the main thread, to ensure synchronisation is now kept between the main thread and the store thread for observers of the `State`.
+
+There are some advantages to this change:
+
+* Simplicity for main-thread `Store`s.
+* Unit testing can now occur on the test framework's thread.
+* Fewer resources needed for context shifting between threads.
```

Comment:
Which configuration are you saying is too complex? 

> the increased likelihood of performance issues appearing from forgetting to switch threads in a Middleware.

This wouldn't increase the likelihood of performance issues unless you choose to change the defaults and not understand the cost of the work you are performing. I would argue that this would help us catch bugs quicker because you would see the cost of doing non-performant work on the wrong dispatcher from the get-go. Today we are hiding this when we use the Store thread to do the wrong kind of work on it. 😄

Reply:
> Which configuration are you saying is too complex?

I didn't mean to imply "too complex", but I do think generally that configurability == complexity. Yes, the complexity is currently hidden by abstractions, but that can be a useful tool for simplifying implementations.

This is a good point, but if we wanted we could pretty easily create profiling middleware that let us understand this with more rich detail rather than trying to catch it in production.

Reply:
> This is a good point, but if we wanted we could pretty easily create profiling middleware that let us understand this with more rich detail rather than trying to catch it in production.

We have profile tooling today already. So we start changing Stores to use this optional behaviour, the developer should be testing their code before landing it.

Reply:
Out of curiosity, do we have profile tooling specifically for Stores and Middleware? I was definitely thinking we could create a middleware that make use of the Firefox Profiler through markers etc, but if that already exists that's great to know

Reply:
No, it's just kotlin-code, so any regular debugger or profiler would work. Out of curiosity, what would a Store/Middleware-specific profile provide?